axelrod/tests/strategies/test_finite_state_machines.py

```diff
@@ -2,13 +2,332 @@
+                      for current_state, input_action, next_state, output_action in transitions}
+        self.assertEqual(get_memory_from_transitions(trans_dict,
+                                                     initial_state=1), 2)
+
+    def test_transient_state(self):
+        """Test a setup where we a transient state (no incoming transitions)
+        goes into a Fortress3 (and D) if the opponent D, and goes into a
+        Cooperator if the opponent C.
+
+        The transient state is state 0.  Fortress3 starts at state 1.  And
+        the Cooperator is state 4.
+        """
+        transitions = (
+            (0, C, 4, C),
+            (0, D, 1, D),
+            (1, C, 1, D),
+            (1, D, 2, D),
+            (2, C, 1, D),
+            (2, D, 3, C),
+            (3, C, 3, C),
+            (3, D, 1, D),
+            (4, C, 4, C),
+            (4, D, 4, C)
+        )
+
+        trans_dict = {(current_state, input_action): (next_state, output_action)
+                      for current_state, input_action, next_state, output_action in transitions}
+
+        # If starting in state 4, then treat like Cooperator
+        self.assertEqual(get_memory_from_transitions(trans_dict,
+                                                     initial_state=4), 0)
+        # Start in state 1, then a Fortress3.
+        self.assertEqual(get_memory_from_transitions(trans_dict,
+                                                     initial_state=1), 2)
+
+    def test_infinite_memory_transient_state(self):
+        """A transient state at 0, which goes into either a Cooperator or a TFT.
+        Because an arbitrarily-long chain of C/C may exist, we would need a
+        infinite memory to determine which state we're in, so that we know how
+        to respond to a D.
+        """
+        transitions = (
+            (0, C, 1, C),
+            (0, D, 2, D),
+            (1, C, 1, C),
+            (1, D, 1, C),
+            (2, C, 2, C),
+            (2, D, 2, D)
+        )
+
+        trans_dict = {(current_state, input_action): (next_state, output_action)
+                      for current_state, input_action, next_state, output_action in transitions}
+
+        self.assertEqual(get_memory_from_transitions(
+                trans_dict, initial_state=0), float("inf"))
+
+        self.assertEqual(get_memory_from_transitions(trans_dict,
+                                                     initial_state=2), 1)
+
+    def test_evolved_fsm_4(self):
+        """This should be infinite memory because the C/D self-loop at state 2
+        and state 3.
+        """
+        transitions = (
+            (0, C, 0, C),
+            (0, D, 2, D),
+            (1, C, 3, D),
+            (1, D, 0, C),
+            (2, C, 2, D),
+            (2, D, 1, C),
+            (3, C, 3, D),
+            (3, D, 1, D),
+        )
+
+        trans_dict = {(current_state, input_action): (next_state, output_action)
```

Comment:
Is it worth defining `def transitions_to_dict(transitions)`?

Reply:
Sure.  Done.